failure in wear sample

[daemon stderr] RobolectricHost SandboxRunner[0] failed: Type androidx.wear.ambient.AmbientLifecycleObserver not present
[daemon stderr] java.lang.TypeNotPresentException: Type androidx.wear.ambient.AmbientLifecycleObserver not present
[daemon stderr] 	at java.base/sun.reflect.annotation.TypeNotPresentExceptionProxy.generateException(TypeNotPresentExceptionProxy.java:47)
[daemon stderr] 	at java.base/sun.reflect.annotation.AnnotationInvocationHandler.invoke(AnnotationInvocationHandler.java:89)
[daemon stderr] 	at jdk.proxy2/jdk.proxy2.$Proxy20.value(Unknown Source)
[daemon stderr] 	at org.robolectric.internal.bytecode.ShadowMap.obtainShadowInfo(ShadowMap.java:161)
[daemon stderr] 	at org.robolectric.internal.bytecode.ShadowMap.obtainShadowInfo(ShadowMap.java:146)
[daemon stderr] 	at org.robolectric.internal.SandboxTestRunner.createClassLoaderConfig(SandboxTestRunner.java:375)
[daemon stderr] 	at org.robolectric.RobolectricTestRunner.createClassLoaderConfig(RobolectricTestRunner.java:192)
[daemon stderr] 	at ee.schimke.composeai.daemon.SandboxHoldingRunner.createClassLoaderConfig(SandboxHoldingRunner.kt:54)
[daemon stderr] 	at org.robolectric.RobolectricTestRunner.getSandbox(RobolectricTestRunner.java:256)
[daemon stderr] 	at org.robolectric.RobolectricTestRunner.getSandbox(RobolectricTestRunner.java:64)
[daemon stderr] 	at org.robolectric.internal.SandboxTestRunner.sandboxGroupingClassBlock(SandboxTestRunner.java:140)
[daemon stderr] 	at org.robolectric.internal.SandboxTestRunner.classBlock(SandboxTestRunner.java:119)
[daemon stderr] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:412)
[daemon stderr] 	at org.robolectric.RobolectricTestRunner.run(RobolectricTestRunner.java:142)
[daemon stderr] 	at org.junit.runners.Suite.runChild(Suite.java:128)
[daemon stderr] 	at org.junit.runners.Suite.runChild(Suite.java:27)
[daemon stderr] 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
[daemon stderr] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
[daemon stderr] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
[daemon stderr] 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
[daemon stderr] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
[daemon stderr] 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
[daemon stderr] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
[daemon stderr] 	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
[daemon stderr] 	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
[daemon stderr] 	at org.junit.runner.JUnitCore.run(JUnitCore.java:105)
[daemon stderr] 	at org.junit.runner.JUnitCore.runClasses(JUnitCore.java:62)
[daemon stderr] 	at org.junit.runner.JUnitCore.runClasses(JUnitCore.java:49)
[daemon stderr] 	at ee.schimke.composeai.daemon.RobolectricHost.runJUnit(RobolectricHost.kt:1112)
[daemon stderr] 	at ee.schimke.composeai.daemon.RobolectricHost.workerThreads$lambda$0$0(RobolectricHost.kt:326)
[daemon stderr] 	at java.base/java.lang.Thread.run(Thread.java:1583)
[daemon stderr] Caused by: java.lang.ClassNotFoundException: androidx.wear.ambient.AmbientLifecycleObserver
[daemon stderr] 	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
[daemon stderr] 	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
[daemon stderr] 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
[daemon stderr] 	at java.base/java.lang.Class.forName0(Native Method)
[daemon stderr] 	at java.base/java.lang.Class.forName(Class.java:536)
[daemon stderr] 	at java.base/java.lang.Class.forName(Class.java:515)
[daemon stderr] 	at java.base/sun.reflect.generics.factory.CoreReflectionFactory.makeNamedType(CoreReflectionFactory.java:114)
[daemon stderr] 	at java.base/sun.reflect.generics.visitor.Reifier.visitClassTypeSignature(Reifier.java:125)
[daemon stderr] 	at java.base/sun.reflect.generics.tree.ClassTypeSignature.accept(ClassTypeSignature.java:49)
[daemon stderr] 	at java.base/sun.reflect.annotation.AnnotationParser.parseSig(AnnotationParser.java:442)
[daemon stderr] 	at java.base/sun.reflect.annotation.AnnotationParser.parseClassValue(AnnotationParser.java:428)
[daemon stderr] 	at java.base/sun.reflect.annotation.AnnotationParser.parseMemberValue(AnnotationParser.java:347)
[daemon stderr] 	at java.base/sun.reflect.annotation.AnnotationParser.parseAnnotation2(AnnotationParser.java:282)
[daemon stderr] 	at java.base/sun.reflect.annotation.AnnotationParser.parseAnnotations2(AnnotationParser.java:121)
[daemon stderr] 	at java.base/sun.reflect.annotation.AnnotationParser.parseAnnotations(AnnotationParser.java:73)
[daemon stderr] 	at java.base/java.lang.Class.createAnnotationData(Class.java:4248)
[daemon stderr] 	at java.base/java.lang.Class.annotationData(Class.java:4237)
[daemon stderr] 	at java.base/java.lang.Class.getAnnotation(Class.java:4120)
[daemon stderr] 	at org.robolectric.internal.bytecode.ShadowMap.obtainShadowInfo(ShadowMap.java:150)
[daemon stderr] 	... 27 more
[refresh] no module — activeFile=<none> (none)

--- a/daemon/android/src/main/kotlin/ee/schimke/composeai/daemon/SandboxHoldingRunner.kt
+++ b/daemon/android/src/main/kotlin/ee/schimke/composeai/daemon/SandboxHoldingRunner.kt
@@ -87,18 +87,23 @@ class SandboxHoldingRunner(testClass: Class<*>) : RobolectricTestRunner(testClas
   }
 
   /**
-   * Conditionally registers [ShadowAmbientLifecycleObserver]. The shadow's
-   * `@Implements(AmbientLifecycleObserver::class)` value forces Robolectric's
-   * [org.robolectric.internal.bytecode.ShadowMap.obtainShadowInfo] to resolve
-   * `androidx.wear.ambient.AmbientLifecycleObserver` via reflection during sandbox bootstrap; on
-   * non-Wear consumers (e.g. `:samples:android`, plain Android apps) that class isn't on the
-   * runtime classpath and the resolution throws `TypeNotPresentException`, killing the daemon
-   * before any render runs (issue: PR #891 regression hit by run-agent-audit-samples.py).
+   * Conditionally registers [ShadowAmbientLifecycleObserver].
    *
-   * Returning the shadow only when wear-ambient is loadable keeps the existing wear path intact
-   * while leaving non-wear consumers untouched. The same gate is mirrored on the engine side in
-   * [ee.schimke.composeai.daemon.RobolectricHost] for `AmbientPreviewOverrideExtension` /
-   * `AmbientInputDispatchObserver` instantiation.
+   * Two interlocking precautions keep non-Wear consumers safe (issue #1244, PR #891 regression hit
+   * by `run-agent-audit-samples.py`):
+   * 1. **The shadow uses `@Implements(className = …)` instead of the `value = …` class-literal
+   *    form**, so Robolectric's [org.robolectric.internal.bytecode.ShadowMap.obtainShadowInfo]
+   *    reads the FQN as a `String` rather than dereferencing a deferred `Class<?>` proxy that the
+   *    JVM annotation parser would resolve via the shadow's defining loader. That deferred
+   *    resolution was the path that threw `TypeNotPresentException` mid-sandbox-bootstrap on Wear
+   *    sample classpaths shipping the shadow next to a stale/mismatched wear AAR.
+   * 2. **This gate still hides the shadow when wear-ambient isn't loadable**, so the shadow class
+   *    — which references `AmbientLifecycleObserver` in field and method-parameter types — is
+   *    only ever returned to Robolectric on classpaths where its symbolic links can be resolved.
+   *
+   * The same gate is mirrored on the engine side in [ee.schimke.composeai.daemon.RobolectricHost]
+   * for `AmbientPreviewOverrideExtension` / `AmbientInputDispatchObserver` instantiation —
+   * those concrete classes call into `AmbientStateController` which directly imports wear API.
    */
   override fun getExtraShadows(method: FrameworkMethod): Array<Class<*>> =
     if (isWearAmbientAvailable(javaClass.classLoader)) {

--- a/data/ambient/connector/src/main/kotlin/ee/schimke/composeai/daemon/ShadowAmbientLifecycleObserver.kt
+++ b/data/ambient/connector/src/main/kotlin/ee/schimke/composeai/daemon/ShadowAmbientLifecycleObserver.kt
@@ -24,8 +24,22 @@ import org.robolectric.annotation.RealObject
  * (`onCreate` / `onResume` / `onPause` / `onDestroy`) are no-ops — the controller drives the
  * ambient transitions directly, and the AOSP impl's lifecycle plumbing (which would have
  * registered with `WearableActivityController`) is intentionally bypassed.
+ *
+ * **Issue #1244 — `@Implements(className = …)` instead of `@Implements(value = …)`.** Robolectric's
+ * `ShadowMap.obtainShadowInfo` reads the shadow's `@Implements` annotation via
+ * `clazz.getAnnotation(Implements.class)`, then queries `className()` first and only dereferences
+ * `value()` when `className()` is empty. The class-literal form (`AmbientLifecycleObserver::class`)
+ * stores a deferred `Class<?>` ref in the annotation proxy; calling `.value()` forces the JVM's
+ * `AnnotationInvocationHandler` to resolve `androidx.wear.ambient.AmbientLifecycleObserver`
+ * against the shadow class's defining loader. On Wear-sample daemon classpaths that ship the
+ * shadow but mismatched-runtime-or-stale wear AAR coordinates this throws
+ * `TypeNotPresentException` mid-sandbox-bootstrap — which the `getExtraShadows` gate in
+ * `SandboxHoldingRunner` can't intercept because the failure is deep inside Robolectric's
+ * iteration loop. Using the FQN string keeps the annotation parseable on any classpath; the
+ * shadow itself is only ever loaded when the wear AAR is present (the
+ * [SandboxHoldingRunner.getExtraShadows] gate is preserved as defence in depth).
  */
-@Implements(AmbientLifecycleObserver::class)
+@Implements(className = "androidx.wear.ambient.AmbientLifecycleObserver")
 class ShadowAmbientLifecycleObserver {
 
   @RealObject @Suppress("unused") private lateinit var realObserver: AmbientLifecycleObserver

--- a/data/ambient/connector/src/test/kotlin/ee/schimke/composeai/daemon/ShadowAmbientLifecycleObserverAnnotationTest.kt
+++ b/data/ambient/connector/src/test/kotlin/ee/schimke/composeai/daemon/ShadowAmbientLifecycleObserverAnnotationTest.kt
@@ -1,0 +1,47 @@
+package ee.schimke.composeai.daemon
+
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNotNull
+import org.junit.Test
+import org.robolectric.annotation.Implements
+
+/**
+ * Pins the `@Implements` annotation form on [ShadowAmbientLifecycleObserver] (issue #1244).
+ *
+ * The shadow must declare `@Implements(className = "androidx.wear.ambient.AmbientLifecycleObserver")`
+ * rather than `@Implements(value = AmbientLifecycleObserver::class)`. Robolectric's
+ * `ShadowMap.obtainShadowInfo` reads `className()` first and only falls back to `value().getName()`
+ * when `className` is empty; the class-literal form stores a deferred `Class<?>` reference in the
+ * annotation proxy, and calling `.value()` forces the JVM's `AnnotationInvocationHandler` to
+ * resolve `androidx.wear.ambient.AmbientLifecycleObserver` against the shadow's defining loader.
+ * On daemon classpaths that ship the shadow next to a mismatched / stale wear AAR (the bug report
+ * in issue #1244 surfaced this against the wear sample) that resolution throws
+ * `TypeNotPresentException` deep inside Robolectric's `createClassLoaderConfig` loop — before
+ * `SandboxHoldingRunner.getExtraShadows`'s gate has a chance to intercept the next sandbox.
+ *
+ * Touching `value()` here makes the regression noisy: if anyone reverts the annotation to the
+ * class-literal form, the deferred proxy resolution kicks in inside this test instead of mid-render
+ * in a daemon's stderr.
+ */
+class ShadowAmbientLifecycleObserverAnnotationTest {
+
+  @Test
+  fun `shadow declares className not value`() {
+    val annotation =
+      ShadowAmbientLifecycleObserver::class.java.getAnnotation(Implements::class.java)
+    assertNotNull("ShadowAmbientLifecycleObserver must carry @Implements", annotation)
+    assertEquals(
+      "ShadowAmbientLifecycleObserver must target AmbientLifecycleObserver by className",
+      "androidx.wear.ambient.AmbientLifecycleObserver",
+      annotation!!.className,
+    )
+    // `value()` must default to `void.class` so the annotation proxy never has to resolve a
+    // deferred Class<?> reference. If the class-literal form sneaks back in, this line throws
+    // `TypeNotPresentException` on any loader without the wear AAR.
+    assertEquals(
+      "@Implements.value() must default to void.class so no deferred class resolution runs",
+      Void.TYPE,
+      annotation.value.java,
+    )
+  }
+}

